Cleanup warnings in Scorpio

This issue is a reminder to cleanup warnings in Scorpio. Earlier PRs cleaned up several warnings but some still remain.

 @amametjanov requested cleanup of the following warnings,

IBM compiler:
```
/autofs/nccs-svm1_home1/azamat/repos/E3SM-tests/externals/scorpio/src/clib/pio_darray_int.c:840:47: warning: address of array 'file->varlist' will always evaluate to 'true' [-Wpointer-bool-conversion]
    pioassert(file && file->iosystem && file->varlist && varids && varids[0] >= 0 &&
                                     ~~ ~~~~~~^~~~~~~

/autofs/nccs-svm1_home1/azamat/repos/E3SM/externals/scorpio/src/clib/pio_darray.c:1551:281: warning: format specifies type 'unsigned int' but the argument has type 'MPI_Datatype' (aka 'struct ompi_datatype_t *') [-Wformat]
                                "Writing variable (%s, varid=%d) to file (%s, ncid=%d) failed. Unable to find a default fillvalue for variable, unsupported variable type (MPI type = %x)", pio_get_vname_from_file(file, varid), varid, pio_get_fname_from_file(file), file->pio_ncid, vtype);
                                                                                                                                                                                      ~~                                                                                                ^~~~~
```

GNU, PGI and IBM:
```
/autofs/nccs-svm1_home1/azamat/repos/E3SM-tests/externals/scorpio/src/clib/pio_sdecomps_regex.cpp:251:7: warning: control may reach end of non-void function [-Wreturn-type]
      }
      ^
```

--- a/src/clib/pio_darray.c
+++ b/src/clib/pio_darray.c
@@ -1556,7 +1556,7 @@ int PIOc_write_darray(int ncid, int varid, int ioid, PIO_Offset arraylen, void *
 #endif /* _NETCDF4 */
             else
                 return pio_err(ios, file, PIO_EBADTYPE, __FILE__, __LINE__,
-                                "Writing variable (%s, varid=%d) to file (%s, ncid=%d) failed. Unable to find a default fillvalue for variable, unsupported variable type (MPI type = %x)", pio_get_vname_from_file(file, varid), varid, pio_get_fname_from_file(file), file->pio_ncid, vtype);
+                                "Writing variable (%s, varid=%d) to file (%s, ncid=%d) failed. Unable to find a default fillvalue for variable, unsupported variable type", pio_get_vname_from_file(file, varid), varid, pio_get_fname_from_file(file), file->pio_ncid);
 
             memcpy((char *)wmb->fillvalue + iodesc->mpitype_size * wmb->num_arrays,
                    fill, iodesc->mpitype_size);

--- a/src/clib/pio_darray_int.c
+++ b/src/clib/pio_darray_int.c
@@ -837,7 +837,7 @@ int write_darray_multi_serial(file_desc_t *file, int nvars, int fndims, const in
     int ierr = PIO_NOERR;              /* Return code. */
 
     /* Check inputs. */
-    pioassert(file && file->iosystem && file->varlist && varids && varids[0] >= 0 &&
+    pioassert(file && file->iosystem && varids && varids[0] >= 0 &&
               varids[0] <= PIO_MAX_VARS && iodesc, "invalid input", __FILE__, __LINE__);
 
     LOG((1, "write_darray_multi_serial nvars = %d fndims = %d iodesc->ndims = %d "

--- a/src/clib/pio_nc.c
+++ b/src/clib/pio_nc.c
@@ -3722,7 +3722,6 @@ int PIOc_copy_att(int incid, int ivarid, const char *name,
   int msg = PIO_MSG_COPY_ATT;
   iosystem_desc_t *ios = NULL;
   file_desc_t *ifile = NULL, *ofile = NULL;
-  int mpierr = MPI_SUCCESS;
   int ierr = PIO_NOERR;
 
   /* Find file corresponding to the input file id */

--- a/src/clib/pio_sdecomps_regex.cpp
+++ b/src/clib/pio_sdecomps_regex.cpp
@@ -245,9 +245,8 @@ namespace PIO_Util{
         else if(type == RIGHT_BRACKET){
           return std::string(")");
         }
-        else{
-          assert(0);
-        }
+        /* FIXME: Use PIO exceptions */
+        throw std::runtime_error("Error converting regex op to string, unsupported regex op type");
       }
 
       std::string SDecomp_regex_op::to_string(void ) const

--- a/src/clib/pioc_support.c
+++ b/src/clib/pioc_support.c
@@ -661,10 +661,15 @@ int pio_err(iosystem_desc_t *ios, file_desc_t *file,
     if ((ret = PIOc_strerror(err_num, err_msg)))
         return ret;
 
-    char uerr_msg[PIO_MAX_NAME + 1];
+    /* PIO_MAX_NAME may not be enough to store the entire user error
+     * message, especially when long filenames are provided by the
+     * user in the message
+     */
+    const int UERR_MSG_MAX_LEN = 8192;
+    char uerr_msg[UERR_MSG_MAX_LEN + 1];
     va_list argp;
     va_start(argp, uerr_msg_fmt);
-    vsnprintf(uerr_msg, PIO_MAX_NAME, uerr_msg_fmt, argp);
+    vsnprintf(uerr_msg, UERR_MSG_MAX_LEN, uerr_msg_fmt, argp);
     va_end(argp);
 
     /* If logging is in use, log an error message. */

